toolkit/tools/imagegen/configuration/packagerepo.go

```diff
@@ -0,0 +1,203 @@
+// available during the installation process if not
+type PackageRepo struct {
+	Name    string `json:"Name"`
+	BaseUrl string `json:"BaseUrl"`
+	Install bool   `json:"Install"`
+}
+
+// UnmarshalJSON Unmarshals a PackageRepo entry
+func (p *PackageRepo) UnmarshalJSON(b []byte) (err error) {
+	// Use an intermediate type which will use the default JSON unmarshal implementation
+	type IntermediateTypePackageRepo PackageRepo
+	err = json.Unmarshal(b, (*IntermediateTypePackageRepo)(p))
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+
+	// Now validate the resulting unmarshaled object
+	err = p.IsValid()
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+	return
+}
+
+// IsValid returns an error if the PackageRepo struct is not valid
+func (p *PackageRepo) IsValid() (err error) {
+	err = p.NameIsValid()
+	if err != nil {
+		return
+	}
+
+	err = p.RepoUrlIsValid()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// NameIsValid returns an error if the package repo name is empty
+func (p *PackageRepo) NameIsValid() (err error) {
+	if strings.TrimSpace(p.Name) == "" {
+		return fmt.Errorf("invalid value for package repo name (%s), name cannot be empty", p.Name)
+	}
+	return
+}
+
+// RepoUrlIsValid returns an error if the package url is invalid
+func (p *PackageRepo) RepoUrlIsValid() (err error) {
+	if strings.TrimSpace(p.BaseUrl) == "" {
+		return fmt.Errorf("invalid value for package repo URL (%s), URL cannot be empty", p.BaseUrl)
+	}
+
+	_, err = url.ParseRequestURI(p.BaseUrl)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func writeAdditionalFieldstoRepoFile(fileName string) (err error) {
+	const (
+		gpgKey    = "gpgkey=file:///etc/pki/rpm-gpg/MICROSOFT-RPM-GPG-KEY file:///etc/pki/rpm-gpg/MICROSOFT-METADATA-GPG-KEY\n"
+		enable    = "enabled=1\n"
+		gpgCheck  = "gpgcheck=0\n"
+		skip      = "skip_if_unavailable=True\n"
+		sslVerify = "sslverify=0\n"
+	)
+
+	additionalFields := gpgKey + enable + gpgCheck + skip + sslVerify
+	err = file.Append(additionalFields, fileName)
+	if err != nil {
+		logger.Log.Errorf("Failed to write additional fields")
+	}
+
+	return
+}
+
+func createCustomRepoFile(fileName string, packageRepo PackageRepo) (err error) {
+	err = file.Create(fileName, 0644)
+	if err != nil {
+		logger.Log.Errorf("Error creating file %s", fileName)
+		return
+	}
+
+	// Write the repo identifier field
+	repoId := "[" + packageRepo.Name + "]\n"
+	err = file.Append(repoId, fileName)
+	if err != nil {
+		logger.Log.Errorf("Failed to write repo identifier %s", repoId)
+		return
+	}
+
+	// Write the repo Name field
+	repoName := "name=" + packageRepo.Name + " $releasever $basearch\n"
+	err = file.Append(repoName, fileName)
+	if err != nil {
+		logger.Log.Errorf("Failed to write repo name %s", repoName)
+		return
+	}
+
+	// Write the repo URL field
+	repoUrl := "baseurl=" + packageRepo.BaseUrl + "\n"
```

Comment:
I'd suggest a combination of `stirngs.Builder` and `fmt.Sprintf` for construction of this file.

Reply:
Fixed